ERCS/erc-7504.md

```diff
@@ -0,0 +1,402 @@
+---
+eip: 7504
+title: Router Proxy Smart Contract
+description: Per-function level upgradeable contracts with rich buildable ABIs
+author: Krishang Nadgauda (@nkrishang)
+discussions-to: https://ethereum-magicians.org/t/erc-7504-dynamic-contracts/15551
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-08-25
+---
+
+## Abstract
+
+This proposal standardizes how 'one-proxy-to-many-implementation' smart contracts should store and expose their ABI. For brevity, we refer to 'one-proxy-to-many-implementation' contracts as dynamic contracts. The proposal presents a scheme for 'rich ABI' for dynamic contracts, going beyond function signatures and their respective implementation contract addresses.
+
+## Motivation
+
+The advantages of building dynamic contracts are well documented in [ERC-2535](./eip-2535.md). The motivation for this proposal goes beyond these advantages.
+
+1. **Comprehensibility**
+
+   As a piece of software, smart contracts have become public-facing backends. Discovering, comprehending or verifying smart contract code on block scanners is second nature to developers, and even users of high-level blockchain applications.
+
+   This proposal embraces this aspect of smart contracts, and introduces an easy to digest pattern for writing dynamic contracts, without excess terminology or complexity.
+
+2. **Client friendliness**
+
+   'Clients' refers to scripts, websites and any such domain for code that calls a smart contract.
+
+   The lifecycle for smart contracts includes someone building a client to interact with it. Building a client — graphical or programatic — to interact with a smart contract requires an ABI, and a well defined interpretation of the ABI to instruct the interaction.
+
+   This proposal focuses on client friendliness. Concretely, this means that it should be straightforward to build clients to interact with dynamic contracts.
+
```

Comment:
Could you go into more detail about why implementers would want to choose ERC-7504 over other similar proxy standards? You mention ERC-2535 above, but don't explicitly say why someone would choose ERC-7504 over Diamonds.